openmmtools/forcefactories.py

```diff
@@ -155,7 +155,11 @@ def restrain_atoms(thermodynamic_state, sampler_state, restrained_atoms, sigma=3
         sampler_state.positions -= centroid * distance_unit
 
     # Create a CustomExternalForce to restrain all atoms.
-    restraint_force = openmm.CustomExternalForce('(K/2)*((x-x0)^2 + (y-y0)^2 + (z-z0)^2)')
+    if thermodynamic_state.system.usesPeriodicBoundaryConditions():
```

Comment:
You can directly use `thermodynamic_state.is_periodic`. This avoids making a copy of the internal standard `System`.

Reply:
Does the `system` property of `thermodynamic_state` really create a *copy* of the internal standard system? If so, this seems *super* confusing, since this would be correcft
```python
system = thermodynamic_state.system
system.addForce(force)
thermodynamic_state.system = system
```
while this would be garbage
```python
thermodynamic_state.system.addForce(force)
```
Do we have to have that behavior?

Reply:
There are two reasons for this
1. If the user modifies directly the system, it could get out of sync with the internal state of the `ThermodynamicState` object (for example by changing a barostat temperature), so only a copy is returned.
2. Internally the `ThermodynamicState` stores only a _standard_ state `System` that is shared across compatible thermodynamic states, not the actual system. If you remember, I had to implement this because creating 30-40 copies of the system object for all the alchemical states was a big performance burden.

The documentation [states it explicitly](https://github.com/choderalab/openmmtools/blob/master/openmmtools/states.py#L290).

Reply:
Oof, OK. I suppose the only thing that will make this better is pure Python versions of System.

Reply:
Yeah. A solution to this could be to add a `ThermodynamicState.add_force()` method to bypass having to make a copy. In the end, the `ThermodynamicState` class is a very fancy `System`. This is the same reason why I implemented the `is_periodic` property, for example.

Reply:
OK, I'll use `._is_periodic`.

Reply:
I'm just very worried about things like `ThermodynamicState` violating the "principle of minimum surprise" in API design. There's no way for someone to know that
```python
thermodynamic_state.system.addForce(force)
```
won't actually work the way it seems it should, for example, without having read and comprehended all the API documentation, and the mention that this is a copy is [really buried](http://openmmtools.readthedocs.io/en/latest/api/generated/openmmtools.states.ThermodynamicState.html#openmmtools.states.ThermodynamicState.system) in there.

Reply:
I agree. It can be confusing, although it's quite standard to return a copy of a mutable object. Keep in mind that the `System` object here is exposed only because there are occasions in which we need it, but I think that in the ideal design the user should not need to access it at all, and `ThermodynamicState` should provide the methods for everything (including `add_force()`).

The only alternative I can see is to create some sort of proxy object that callbacks the `ThermodynamicState` every time the system is modified, but at this point, we might as well make `ThermodynamicState` itself the proxy object.